File: workflow-orchestrator/src/main/java/com/omnibus/workflow/application/RulePublicationWorkflowService.java
package com.omnibus.workflow.application;

import com.omnibus.workflow.domain.RulePublicationRequest;
import com.omnibus.workflow.domain.RulePublicationSubmission;
import io.camunda.client.CamundaClient;
import io.camunda.client.api.response.ProcessInstanceEvent;
import java.time.Instant;
import java.util.HashMap;
import java.util.Map;
import org.springframework.stereotype.Service;
import reactor.core.publisher.Mono;
import reactor.core.scheduler.Schedulers;

@Service
public class RulePublicationWorkflowService {

    static final String PROCESS_ID = "rule-publication-process";

    private final CamundaClient camundaClient;

    public RulePublicationWorkflowService(CamundaClient camundaClient) {
        this.camundaClient = camundaClient;
    }

    public Mono<RulePublicationSubmission> startPublication(RulePublicationRequest request) {
        return Mono.fromCallable(() -> {
                ProcessInstanceEvent event = camundaClient.newCreateInstanceCommand()
                    .bpmnProcessId(PROCESS_ID)
                    .latestVersion()
                    .variables(buildVariables(request))
                    .send()
                    .join();

                return new RulePublicationSubmission(
                    Long.toString(event.getProcessInstanceKey()),
                    PROCESS_ID,
                    "STARTED",
                    request.name(),
                    request.requestedBy()
                );
            })
            .subscribeOn(Schedulers.boundedElastic());
    }

    private Map<String, Object> buildVariables(RulePublicationRequest request) {
        Map<String, Object> variables = new HashMap<>();
        variables.put("ruleName", request.name());
        variables.put("ruleContent", request.content());
        variables.put("requestedBy", request.requestedBy());
        variables.put("requestedAt", Instant.now().toString());
        return variables;
    }
}

File: services/core/workflow-orchestrator/src/main/kotlin/pl/writeonly/omnibus/workflow/application/RulePublicationWorkflowService.kt
package pl.writeonly.omnibus.workflow.application

import io.camunda.client.CamundaClient
import io.camunda.client.api.response.ProcessInstanceEvent
import org.springframework.stereotype.Service
import pl.writeonly.omnibus.workflow.domain.RulePublicationRequest
import pl.writeonly.omnibus.workflow.domain.RulePublicationSubmission
import java.time.Instant

@Service
class RulePublicationWorkflowService(
    private val camundaClient: CamundaClient,
) {
    fun startPublication(request: RulePublicationRequest): RulePublicationSubmission {
        val event: ProcessInstanceEvent =
            camundaClient.newCreateInstanceCommand()
                .bpmnProcessId(PROCESS_ID)
                .latestVersion()
                .variables(buildVariables(request))
                .send()
                .join()

        return RulePublicationSubmission(
            event.processInstanceKey.toString(),
            PROCESS_ID,
            "STARTED",
            request.name,
            request.requestedBy,
        )
    }

    private fun buildVariables(request: RulePublicationRequest): Map<String, Any> =
        hashMapOf(
            "ruleName" to request.name,
            "ruleContent" to request.content,
            "requestedBy" to request.requestedBy,
            "requestedAt" to Instant.now().toString(),
        )

    companion object {
        const val PROCESS_ID: String = "rule-publication-process"
    }
}
